modules/parachains/src/lib.rs

```diff
@@ -405,9 +408,23 @@ pub mod pallet {
 								.saturating_sub(WeightInfoOf::<T, I>::parachain_head_pruning_weight(T::DbWeight::get()));
 						}
 					}
+
+					// even though we may have accepted some parachain heads, we can't allow relayers to submit
+					// proof with unused trie nodes
+					// => treat this as an error
+					//
+					// (we can throw error here, because now all our calls are transactional)
+					storage.ensure_no_unused_nodes()
 				},
 			)
-			.map_err(|_| Error::<T, I>::InvalidStorageProof)?;
+			.map_err(|e| {
+				log::trace!(target: LOG_TARGET, "Failed to read parachain heads. Storage proof is invalid: {:?}", e);
+				Error::<T, I>::InvalidStorageProof
+			})?
+			.map_err(|e| {
+				log::trace!(target: LOG_TARGET, "Failed to read parachain heads: {:?}", e);
+				Error::<T, I>::InvalidStorageProof
+			})?;
```

Comment:
Nit: I think they might be reversed. I think the second one would be the one returned by `ensure_no_unused_nodes` which is `Storage proof is invalid` for sure. And the first one is the one returned by `parse_finalized_storage_proof` which could also be `UnknownHeader` for example. Not sure if this was the intention. 

Anyway, it's good we have #1881 . It will help us with this.

Also maybe we don't need to log `Failed to read parachain heads.` for the second one. Technically the parachain heads could be read, also some update might have taken place, but the proof was invalid and an error was thrown at the end.